### DISCUSSION TITLE
TypeScript types for `watch` callback's `name`

### DISCUSSION THREAD

**Discussion by joshkel** [react-hook-form/react-hook-form#6392] [Q&A] [answered]
I'm having a few problems / questions with the change from #6171 to provide a type for the `watch` callback's `name` parameter:

First, I see that `name` is now permitted to be an array. Under what circumstances would it be an array? I checked the source code, and it looks like it's always a string. (Of course, I could easily be missing something; or maybe array use is coming in a future release?)

Second, to accommodate the possibly of array, I added an `!Array.isArray(name)` type guard in my callback. However, that doesn't work: `name` may be `FieldPath<TFieldValues>[] | readonly FieldPath<TFieldValues>[]`, and the `Array.isArray` type guard only excludes the `FieldPath<TFieldValues>[]` possibility. (See [this TypeScript playground](https://www.typescriptlang.org/play?ssl=1&ssc=29&pln=1&pc=49#code/GYVwdgxgLglg9mABFApgZygCgA4EMBOuAtgFyIb4xgDmiAPovirgCYIA2AnuVJTQNoBdejz7UhASkQBvALAAoRIhjBEmAIQBBfIU4A6GGm26cBYhKlzFSxBARo47FHvZxqpwkT1Q4AGTgQuE4AqtjYKPgAwrhoKJgWANwKSgC+CilAA) for a demonstration.) This looks like it may be a bug in TypeScript (it's discussed [here](https://github.com/microsoft/TypeScript/issues/17002#issuecomment-516013923)), but I wondered if it would make sense for React Hook Form to accommodate by removing the `readonly FieldPath[]` option.

Third, and related, I don't understand why `name` is declared as `FieldPath<TFieldValues>[] | readonly FieldPath<TFieldValues>[]`. `readonly` only has meaning within TypeScript's type system, so it has no meaning at runtime, and there's no way for a function to check whether a value is `readonly` or not. I think that the type declaration means "it's either an array that the callback is expected to potentially modify or an array that the callback is not expected to modify"; it seems like it should either be one or the other. If I'm misunderstanding or missing something, please let me know.

**Comment by bluebill1049** [comment, marked as answer]
> First, I see that name is now permitted to be an array. Under what circumstances would it be an array?

when users want to watch multiple inputs.

**Reply by joshkel** [reply]
Apologies if I was unclear; I'm asking specifically about the callback form of `watch`. As I understand it, the callback is invoked once per changed field; there isn't an option of saying up front how many inputs the user wants to watch.

**Reply by bluebill1049** [reply]
That's correct, if this is what you refer to 

```ts
watch((data) => {
  console.log(data)
})
```

**Reply by joshkel** [reply]
Yes, I'm referring to

```ts
watch((data) => {
  console.log(data)
})
```

Thanks.

Does that mean that the type definitions should be changed to

```ts
export type WatchObserver<TFieldValues> = (
  value: UnpackNestedValue<TFieldValues>,
  info: {
    name?: FieldPath<TFieldValues>;
    type?: EventType;
    value?: unknown;
  },
);
```

?

**Reply by bluebill1049** [reply]
you are right! please send us a PR 👍

**Reply by joshkel** [reply]
For anyone else who comes across this, the PR is at https://github.com/react-hook-form/react-hook-form/pull/6404. Thanks for your help!
